general/jit/parser.cpp

```diff
@@ -0,0 +1,580 @@
+         check(is_space() || is_id() || is_star() || is_amp() ||
+               is_eq() || is_lt() || is_gt() || is_coma(),
+               "while parsing the arguments.");
+         if (is_lt()) { lt++; }
+         if (is_gt()) { lt--; }
+         if (is_space() && !id.empty() && id.back() != '.' && !ker.is_eq)
+         {
+            if (last(id) == "MFEM_JIT") { ker.mv_to_targs = true; id = head(id); }
+            id += '.';
+         }
+         if (is_eq())
+         {
+            ker.is_eq = true;
+            if (id.back() == '.') { most(id); }
+            if (ker.Targs.find(last(id)) == string::npos)
+            { check(false, string("Could not find T_")+to_upper(last(id))); }
+         }
+         if (is_id() && !ker.is_eq) { id += in.peek(); }
+         if (is_coma() && lt == 0)
+         {
+            if (id.back() == '.') { most(id); }
+            add_id(), id.clear(), ker.is_eq = false, ker.mv_to_targs = false;
+         }
+         put();
+      }
+      add_id();
+      check(!ker.Tparams.empty(), "MFEM_JIT function without JIT argument!");
+      if (!ker.is_templated) { ker.Tparams += " bool done = true"; }
+
+      ker.advance(); // Params => Prefix
+
+      // Make sure we hit the last ')' of the arguments
+      check(put()==')',"no last ')' in kernel");
+
+      // Make sure we are about to start a compound statement
+      next_check([&]() { return put()=='{';}, "no compound statement found");
+
+      // Generate the kernel prefix for this kernel
+      ker.dup.clear(), ker.dup.str(string());
+      ker.src.clear(), ker.src.str(string());
+      ker.advance(); // Prefix => Body
+
+      ker.src << "\n\tconst char *source = R\"_(";
+      // mfem, forall & jit includes are added from command line from jit.cpp
+      ker.src << "\nusing namespace mfem;";
+      for (auto inc: ker.includes) { ker.src << "\n#include \"" + inc +"\""; }
+      ker.src << "\n\ntemplate<" << ker.Tparams << ">";
+      ker.src << "\nvoid " << ker.name << "_%016lx"
+              << "(" << ker.Sparams0 << "){";
+      ker.src << pp_line();
+      block = 0; // Start counting the block statements
+   }
+
+   /*
+    * mfem_jit_postfix prepare:
+    *   - the JIT inputs: compiler, flags, libraries from the build system,
+    *   - computes the hash of the source, compiler, libraries, flags,
+    *     MFEM_SOURCE_DIR and MFEM_INSTALL_DIR,
+    *   - the per-kernel local unordered_map which holds the different kernel
+    *     symbols that have already been loaded,
+    *   - updates the pre-processor line.
+    */
+   void mfem_jit_postfix() // output all kernel source, with updated hash
+   {
+      ker.src << "}\nextern \"C\" void k%016lx"
+              << "(" << ker.Sparams0 << "){"
+              << "\n\t" << ker.name << "_%016lx<" << ker.Tformat << ">"
+              << "("<< ker.Sargs_us << ");";
+
+      // these settings are global to MFEM
+      const char *cxx = "" MFEM_CXX;
+      const char *libs = "" MFEM_EXT_LIBS;
```

Comment:
I had to update this to get `./ex1 -pa` to work:
```suggestion
      const char *libs = "-L" MFEM_INSTALL_DIR "/lib -lmfem " MFEM_EXT_LIBS;
```

Reply:
Just to elaborate: this change was when I had `MFEM_DEBUG` enabled -- one of the JIT compilations could not find `mfem::mfem_error` which was needed by the `MFEM_ASSERT`s in `linalg/dtensor.hpp`.

Reply:
`done`